Library/Homebrew/formula.rb

```diff
@@ -2429,34 +2407,149 @@ def to_hash_with_variations
+        "using"  => (T.must(head).using if T.must(head).using.is_a?(Symbol)),
+      }
+    end
+
+    hash
+  end
+
+  # @private
+  def serialized_requirements
+    requirements = self.class.spec_syms.to_h do |sym|
+      [sym, send(sym)&.requirements]
+    end
+
+    merge_spec_dependables(requirements).map do |data|
+      req = data[:dependable]
+      req_name = req.name.dup
+      req_name.prepend("maximum_") if req.respond_to?(:comparator) && req.comparator == "<="
+      req_version = if req.respond_to?(:version)
+        req.version
+      elsif req.respond_to?(:arch)
+        req.arch
+      end
+      {
+        "name"     => req_name,
+        "cask"     => req.cask,
+        "download" => req.download,
+        "version"  => req_version,
+        "contexts" => req.tags,
+        "specs"    => data[:specs],
       }
     end
+  end
+
+  # @private
+  def dependencies_hash
+    # Create a hash of spec names (stable/head) to the list of dependencies under each
+    dependencies = self.class.spec_syms.to_h do |sym|
+      [sym, send(sym)&.declared_deps]
+    end
+    dependencies.transform_values! { |deps| deps&.reject(&:implicit?) } # Remove all implicit deps from all lists
+
+    hash = {}
+
+    dependencies.each do |spec_sym, spec_deps|
+      next if spec_deps.nil?
+
+      dep_hash = if spec_sym == :stable
+        hash
+      else
+        next if spec_deps == dependencies[:stable]
+
+        hash["#{spec_sym}_dependencies"] ||= {}
+      end
+
+      dep_hash["build_dependencies"] = spec_deps.select(&:build?)
+                                                .reject(&:uses_from_macos?)
+                                                .map(&:name)
+                                                .uniq
+      dep_hash["dependencies"] = spec_deps.reject(&:optional?)
+                                          .reject(&:recommended?)
+                                          .reject(&:build?)
+                                          .reject(&:test?)
+                                          .reject(&:uses_from_macos?)
+                                          .map(&:name)
+                                          .uniq
+      dep_hash["test_dependencies"] = spec_deps.select(&:test?)
+                                               .reject(&:uses_from_macos?)
+                                               .map(&:name)
+                                               .uniq
+      dep_hash["recommended_dependencies"] = spec_deps.select(&:recommended?)
+                                                      .reject(&:uses_from_macos?)
+                                                      .map(&:name)
+                                                      .uniq
+      dep_hash["optional_dependencies"] = spec_deps.select(&:optional?)
+                                                   .reject(&:uses_from_macos?)
+                                                   .map(&:name)
+                                                   .uniq
```

Comment:
@Bo98 you've historically had opinions about these: can you share here? Something like having the tags for the dependencies rather than duplicating.

Reply:
OK with this iterating this being in a second PR (since all this will be not-finalised under an env flag) but yeah I will try get something together on Wednesday.

Reply:
I think it should be possible to see the difference either way and I'm happy with waiting. It might be easier to review in a separate PR though.

I wonder if this format would make sense too.

```rb
"dependencies": {
  "a": ["build", "test"]
}
```

The current file size is around 10MB.

Reply:
We also wouldn't need to share that logic between both hash functions anymore.

Reply:
> I wonder if this format would make sense too.

Depends if we want to combine `head_dependencies` and `uses_from_macos` into it or not, e.g.

```
"dependencies": [
  {
    "name": "a",
    "tags": ["build", "test"]
    "specs": ["stable", "head"].
    "uses_from_macos": { "since": "catalina" }
  }
]
```

Reply:
> Earliest I can get something together is Wednesday.

No rush here 👍🏻 



> Motivation being it will reduce the logic massively on both the generation and reading side as we don't need to do a mess of `select`, `reject` etc and reading and combining several different arrays. `uses_from_macos` and `uses_from_macos_bounds` would also fold into this. The code has always looked horrible to me.

This makes a lot of sense to me 👍🏻 



> Specs combining and separation would therefore require a little bit of logic, but doing so could significantly reduce the overhead of new format in the JSON itself.

Seems nice if the reading logic doesn't blow up too much as a result 👍🏻